src/v/kafka/server/group_manager.cc

```diff
@@ -222,38 +222,47 @@ ss::future<> group_manager::handle_partition_leader_change(
          * we just became leader. make sure the log is up-to-date. see
          * struct group_log_record_key{} for more details.
          */
-        return inject_noop(p->partition, timeout).then([this, timeout, p] {
-            /*
-             * the full log is read and deduplicated. the dedupe processing is
-             * based on the record keys, so this code should be ready to
-             * transparently take advantage of key-based compaction in the
-             * future.
-             */
-            storage::log_reader_config reader_config(
-              p->partition->start_offset(),
-              model::model_limits<model::offset>::max(),
-              0,
-              std::numeric_limits<size_t>::max(),
-              kafka_read_priority(),
-              raft::data_batch_type,
-              std::nullopt,
-              std::nullopt);
-
-            return p->partition->make_reader(reader_config)
-              .then([this, p, timeout](model::record_batch_reader reader) {
-                  return std::move(reader)
-                    .consume(recovery_batch_consumer(p->as), timeout)
-                    .then([this, p](recovery_batch_consumer_state state) {
-                        // avoid trying to recover if we stopped the reader
-                        // because an abort was requested
-                        if (p->as.abort_requested()) {
-                            return ss::make_ready_future<>();
-                        }
-                        return recover_partition(p->partition, std::move(state))
-                          .then([p] { p->loading = false; });
-                    });
-              });
-        });
+
+        return _catchup_lock.take_writer_lock()
+          .then([this, timeout, p] {
+              return inject_noop(p->partition, timeout)
+                .then([this, timeout, p] {
+                    /*
+                     * the full log is read and deduplicated. the dedupe
+                     * processing is based on the record keys, so this code
+                     * should be ready to transparently take advantage of
+                     * key-based compaction in the future.
```

Comment:
@dotnwat, since we added proper x-segment compaction, should we simply move this log to a compacted log now?

Reply:
yes but we should wait for roko's mirror maker experiments. there is some thought that maybe we need to make a consumer group topic that is compatible with the format of the kafka __consumer_offsets topic because some tools and maybe mirror maker interact with this topic directly even thought it is "internal". in this particular case we'd need to figure out a plan for migrating data and users to the new topic.

Reply:
it's already compacted, we create consumer group topic in find_coordinator.cc and set `cleanup_policy_bitflags` to `model::cleanup_policy_bitflags::compaction`